src/chainstate/stacks/db/blocks.rs

```diff
@@ -3340,8 +3348,51 @@ impl StacksChainState {
         }
 
         // 4: the account nonces must be correct
-        let (origin, payer) = StacksChainState::check_transaction_nonces(clarity_connection, &tx)
-            .map_err(|e| MemPoolRejection::BadNonces(e))?;
+        let (origin, payer) = match StacksChainState::check_transaction_nonces(clarity_connection, &tx) {
+            Ok(x) => x,
+            Err((mut e, (origin, payer))) => {
+                // let's see if the tx has matching nonce in the mempool
+                //  ->  you can _only_ replace-by-fee or replace-across-fork
+                //      for nonces that increment the current chainstate's nonce.
+                //      if there are "chained" transactions in the mempool,
+                //      this check won't admit a rbf/raf for those.
+                let origin_addr = tx.origin_address();
+                let origin_nonce = tx.get_origin().nonce();
+                let origin_next_nonce = MemPoolDB::get_next_nonce_for_address(mempool, &origin_addr)?;
+                if origin_next_nonce < origin.nonce {
+                    return Err(e.into())
+                }
+                if origin_next_nonce - origin.nonce >= MAXIMUM_MEMPOOL_TX_CHAINING {
+                    return Err(MemPoolRejection::TooMuchChaining)
+                }
+                if origin_nonce != origin_next_nonce {
+                    e.is_origin = true;
+                    e.principal = origin_addr.into();
+                    e.expected = origin_next_nonce;
+                    e.actual = origin_nonce;
+                    return Err(e.into())
+                }
```

Comment:
The timing is definitely not right for introducing this feature at this point, but what's our take in general on being able to receive transactions, out of nonce order (with some kind of limit)?

Reply:
I'd be open to it -- it is already the case that we cannot perform reliable "balance" checks on chained transactions, so I don't think out of order nonce arrivals would have any new adverse affects. Is this mainly useful for tooling that may broadcast a few transactions at once, but not necessarily order them correctly?

Reply:
It would be useful in general to support this because transaction messages might get lost or reordered as they are propagated through the p2p network.  Balance checks would only work off of the highest contiguous sequence of transactions.

Reply:
However, storing and resolving out-of-order transactions doesn't have to happen in the mempool.  For example, an alternative approach could be to have each node simply buffer the last X transactions that have arrived, and serve them to neighbors who ask for them.  Then, if a node gets a transaction from a neighbor with a higher-than-expected nonce, the node can ask the neighbor for the missing transaction(s).